experimental/web/sample_webgpu/main.c

```diff
@@ -279,8 +293,192 @@ static iree_status_t parse_inputs_into_call(
+  iree_hal_buffer_view_t* buffer_view = NULL;
+  iree_hal_dim_t buffer_view_shape[10];
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_buffer_view_shape(userdata->buffer_view,
+                                        /*rank_capacity=*/10, buffer_view_shape,
+                                        /*out_shape_rank=*/NULL);
+  }
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_buffer_view_create(
+        heap_buffer, iree_hal_buffer_view_shape_rank(userdata->buffer_view),
+        buffer_view_shape,
+        iree_hal_buffer_view_element_type(userdata->buffer_view),
+        iree_hal_buffer_view_encoding_type(userdata->buffer_view),
+        iree_allocator_system(), &buffer_view);
+  }
+
+  if (iree_status_is_ok(status)) {
+    fprintf(stdout, "Call output:\n");
+    status = iree_hal_buffer_view_fprint(stdout, buffer_view,
+                                         /*max_element_count=*/4096,
+                                         iree_allocator_system());
+    fprintf(stdout, "\n");
+  }
+  iree_hal_buffer_view_release(buffer_view);
+  iree_hal_buffer_release(heap_buffer);
+
+  if (!iree_status_is_ok(status)) {
+    fprintf(stderr, "buffer_map_sync_callback error:\n");
+    iree_status_fprint(stderr, status);
+    iree_status_free(status);
+  }
+
+  iree_allocator_free(iree_allocator_system(), data_cloned);
+
+  iree_allocator_free(iree_allocator_system(), userdata);
+}
+
+static iree_status_t print_buffer_view(iree_hal_device_t* device,
+                                       iree_hal_buffer_view_t* buffer_view,
+                                       iree_hal_allocator_t* heap_allocator) {
+  iree_status_t status = iree_ok_status();
+
+  iree_device_size_t data_length =
+      iree_hal_buffer_view_byte_length(buffer_view);
+  iree_hal_buffer_t* buffer = iree_hal_buffer_view_buffer(buffer_view);
+
+  // ----------------------------------------------
+  // Transfer from device memory to mappable host memory.
+  iree_hal_buffer_t* mappable_buffer = NULL;
+  iree_device_size_t target_offset = 0;
+  const iree_hal_buffer_params_t target_params = {
+      .type =
+          IREE_HAL_MEMORY_TYPE_HOST_LOCAL | IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE,
+      .usage = IREE_HAL_BUFFER_USAGE_TRANSFER | IREE_HAL_BUFFER_USAGE_MAPPING,
+  };
+  status = iree_hal_allocator_allocate_buffer(
+      iree_hal_device_allocator(device), target_params, data_length,
+      iree_const_byte_span_empty(), &mappable_buffer);
+  // Issue synchronous device copy.
+  if (iree_status_is_ok(status)) {
+    const iree_hal_transfer_command_t transfer_command = {
+        .type = IREE_HAL_TRANSFER_COMMAND_TYPE_COPY,
+        .copy =
+            {
+                .source_buffer = buffer,
+                .source_offset = 0,
+                .target_buffer = mappable_buffer,
+                .target_offset = target_offset,
+                .length = data_length,
+            },
+    };
+    status = iree_hal_device_transfer_and_wait(
```

Comment:
I'm not sure how this works as this is a blocking call (`and_wait`)?
since you aren't using the feature of the function I'd suggest just inlining it - it's just calling iree_hal_create_transfer_command_buffer and iree_hal_device_queue_execute and you're going to need control over the wait in order to be non-blocking (it calls iree_hal_semaphore_wait, which is a no-go) - you want to get a wait source back and pass it to iree_loop_wait_one.

Reply:
Okay, inlined it and added a TODO for wait source + iree_loop_wait_one (thanks for spelling that out, I'll get to it when I add the Emscripten loop impl :D)